cmd/k8s-local-pv-provisioner/run.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/apex/log"
+
+	gocli "gopkg.in/src-d/go-cli.v0"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func init() {
+	app.AddCommand(&RunCommand{})
+}
+
+type RunCommand struct {
+	gocli.PlainCommand `name:"run" short-description:"run a watcher for PVs" long-description:"Run an in-cluster watcher for PVs and create the needed paths if needed"`
+	NodeName           string `long:"node-name" required:"true" env:"NODE_NAME" description:"Hostname of the current node the pod runs on"`
+	KubernetesContext  string `long:"context" env:"KUBERNETES_CONTEXT" description:"If set the program will oad the kubernetes configuraion from a kubeconfig file for the given context"`
+	RootfsPath         string `long:"rootfs-path" env:"ROOTFS_PATH" default:"/rootfs" description:"Path to the mounted root file system of the node"`
+}
+
+func (r *RunCommand) ExecuteContext(ctx context.Context, args []string) error {
+	client, err := r.getClientSet()
+	if err != nil {
+		return err
+	}
+
+	watcher, err := client.CoreV1().PersistentVolumes().Watch(meta_v1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	watcher.ResultChan()
+L:
+	for {
+		select {
+		case <-ctx.Done():
+			watcher.Stop()
+			break L
+		case event := <-watcher.ResultChan():
+			if event.Type == watch.Error {
+				return errors.New("Got kubernetes watch error")
+			}
+			if event.Type == watch.Added || event.Type == watch.Modified {
+				r.setUpPV(event.Object.(*core_v1.PersistentVolume))
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *RunCommand) getClientSet() (*kubernetes.Clientset, error) {
+	var config *rest.Config
+	var err error
+	if r.KubernetesContext != "" {
+		config, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			clientcmd.NewDefaultClientConfigLoadingRules(),
+			&clientcmd.ConfigOverrides{
+				CurrentContext: r.KubernetesContext,
+			},
+		).ClientConfig()
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return kubernetes.NewForConfig(config)
+}
+
+func (r *RunCommand) setUpPV(pv *core_v1.PersistentVolume) {
+	if pv.Spec.Local == nil {
+		log.Infof("%s is not a local volume, skipping", pv.ObjectMeta.Name)
+		return
+	}
+
+	matches := false
+	// We can assume this is not nil as `Spec.Local` is invalid without NodeAffinity
+	for _, term := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for _, matchExpression := range term.MatchExpressions {
+			if matchExpression.Key != "kubernetes.io/hostname" { // we only have hostname support
+				continue
+			}
+			if matchExpression.Operator == core_v1.NodeSelectorOperator("In") {
+				for _, value := range matchExpression.Values {
+					if value == r.NodeName { // replace me with an env var!!!!
```

Comment:
I don't fully understand this comment. How would a variable avoid the fact of checking that this executable should only care about those PVs related to its own NodeName?

Reply:
Or is it related to the fact that you can only have a reliable node name in hostPath?

Reply:
Ok, taking a look to https://github.com/src-d/charts/pull/130/files it definitely looks the last option, so please go that route

Reply:
A lost development comment :disappointed: I did replace it but forgot to remove the note to myself

Reply:
Hmm i heard that before, need to check if I can get that in VSCode
(a terrible slippy style lookout analyzer idea came to mind... but let's not go there)